documentation/public/user-guide/rust-crate/how-to-parse-a-file/index.md

```diff
@@ -1,3 +1,304 @@
+use slang_solidity::cst::Node;
+use slang_solidity::kinds::{RuleKind, TokenKind};
+use slang_solidity::language::Language;
+
+fn main() -> Result<()> {
+  let language = Language::new(Version::parse("0.8.0")?)?;
+  // Or, using the version directly:
+  let language = Language::new(Version::new(0, 8, 0))?;
+
+  // ...
+}
+```
+
+Now, we need to read the source code. The `parse` function accepts a `&str` slice, so we need to load it from the file directly or slice from an existing `String` buffer.
+
+Let's use the convenient [`std::fs::read_to_string`](https://doc.rust-lang.org/stable/std/fs/fn.read_to_string.html) helper to make our lives a bit easier.
+
+```rust
+// ...cont.
+let source_code = std::fs::read_to_string("file.sol")?;
+
+let parse_tree/*: ParseOutput*/ = language.parse(RuleKind::SourceUnit, &source_code);
+```
+
+The resulting `ParseOutput` type exposes these helpful functions:
+
+-   `fn errors()/is_valid()` that return structured parse errors, if any,
+-   `fn tree()` that gives us back a CST (partial if there were parse errors),
+-   `fn create_tree_cursor()` that creates a `Cursor` type used to conveniently walk the parse tree.
+
+### Example 1: Reconstruct the Solidity file
+
+Let's try the same example, only now using the API directly.
+
+We'll start with this file:
+
+```solidity
+// file: file.sol
+pragma solidity ^0.8.0;
+```
+
+#### Step 1: Parse the Solidity file
+
+We can re-use the snippets above:
+
+```rust
+// ...cont.
+let source_code = std::fs::read_to_string("file.sol")?;
+
+let parse_tree/*: ParseOutput*/ = language.parse(RuleKind::SourceUnit, &source_code);
+```
+
+#### Step 2: Reconstruct the source code
+
+The API already includes a convenience `Node::unparse` that does exactly that:
+
+```rust
+let output/*: String*/ = parse_tree.node().unparse();
+assert_eq!(output, "pragma solidity ^0.8.0\n");
+```
+
+However, let's do that ourselves to exercise the tree walking.
+
+The `Cursor` type implements an `Iterator` trait by yielding the tree nodes in a depth-first search (DFS) fashion.
+
+To only visit the `Token` nodes, we can use the [`Iterator::filter_map`](https://doc.rust-lang.org/stable/std/iter/trait.Iterator.html#method.filter_map) iterator adapter that filters and maps the tree token nodes using the helper [`Node::into_token`](https://docs.rs/slang_solidity/latest/slang_solidity/cst/enum.Node.html#method.into_token) conversion function.
+
+```rust
+let cursor = parse_tree.create_tree_cursor();
+let only_tokens_iter = cursor.filter_map(|node| node.into_token());
+
+let mut source = vec![];
+for token_node in only_tokens_iter {
+  source.push_str(&token_node.text);
+}
+
+assert_eq!(output, "pragma solidity ^0.8.0\n");
+```
+
+### Example 2: List the top-level contracts and their names
+
+In addition to the `Iterator` implementation, the `Cursor` type also exposes procedural-style functions that allow you to navigate the source in an imperative fashion:
+
+```rust
+const SOURCE: &'static str = "
```

Comment:
I suggest writing these as unit tests under `crates/solidity/outputs/npm/tests/src/doc-examples/` and embed them here, so that we can verify them using `cargo test`.

Reply:
Changed almost all of these to snippets of tested code in https://github.com/NomicFoundation/slang/pull/716/commits/d05e7577a022a362bb580554e12c313720876073 and https://github.com/NomicFoundation/slang/pull/716/commits/3b1b6f67d7270b5703df1844e383fb2a58c7eeae